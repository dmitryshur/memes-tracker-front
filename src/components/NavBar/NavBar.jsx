import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { toggleNavbar } from "../../actions/visuals/visuals";
import "./NavBar.scss";

export const NavBar = props => {
  const handleButtonClick = event => {
    event.stopPropagation();
    props.toggleNavbar();
  };

  // close the navbar
  const handleOverlayClick = event => {
    event.stopPropagation();
    props.toggleNavbar();
  };

  return (
    <React.Fragment>
      <div
        className="cn-button"
        id="cn-button"
        onClick={handleButtonClick}
        onKeyPress={handleButtonClick}
        role="button"
        tabIndex="0"
      >
        {props.navbarStatus === "closed" ? "+" : "-"}
      </div>
      <div
        id="cn-wrapper"
        className={props.navbarStatus === "closed" ? "cn-wrapper" : "cn-wrapper opened-nav"}
      >
        <ul>
          <li>
            <a href="#">
              <span className="fas fa-user-plus" />
            </a>
          </li>
          <li>
            <a href="#">
              <span className="fas fa-home" />
            </a>
          </li>
          <li>
            <a href="#">
              <span className="fas fa-star" />
            </a>
          </li>
        </ul>
      </div>
      <div
        id="cn-overlay"
        className={props.navbarStatus === "closed" ? "cn-overlay" : "cn-overlay on-overlay"}
        onClick={handleOverlayClick}
        onKeyPress={handleOverlayClick}
        role="button"
        tabIndex="0"
      />
    </React.Fragment>
  );
};

NavBar.propTypes = {
  navbarStatus: PropTypes.string,
  toggleNavbar: PropTypes.func
};

NavBar.defaultProps = {
  navbarStatus: "closed",
  toggleNavbar: () => {}
};

const mapDispatchToProps = dispatch => ({
  toggleNavbar: section => dispatch(toggleNavbar(section))
});

export default connect(undefined, mapDispatchToProps)(NavBar);
