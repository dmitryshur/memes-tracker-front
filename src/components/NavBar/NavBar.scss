//@import url(http://netdna.bootstrapcdn.com/font-awesome/2.0/css/font-awesome.css);
@import url(https://use.fontawesome.com/releases/v5.0.10/css/all.css);

* {
  -moz-box-sizing: border-box;
  box-sizing: border-box;
  margin: 0;
  padding: 0;
  list-style: none;
  position: relative;
}

.cn-wrapper {
  -moz-transform: scale(0.1);
  -moz-transition: all 0.3s ease;
  -ms-transform: scale(0.1);
  -webkit-transform: scale(0.1);
  -webkit-transition: all 0.3s ease;
  border-radius: 50%;
  bottom: -13em;
  font-size: 3em;
  height: 26em;
  left: 50%;
  margin-left: -13em;
  overflow: hidden;
  pointer-events: none;
  position: fixed;
  transform: scale(0.1);
  transition: all 0.3s ease;
  width: 26em;
  z-index: 10;

  li {
    -moz-transform-origin: 100% 100%;
    -moz-transition: border 0.3s ease;
    -ms-transform-origin: 100% 100%;
    -webkit-transform-origin: 100% 100%;
    -webkit-transition: border 0.3s ease;
    font-size: 1.5em;
    height: 10em;
    left: 50%;
    margin-left: -10em;
    margin-top: -1.3em;
    overflow: hidden;
    position: absolute;
    top: 50%;
    transform-origin: 100% 100%;
    transition: border 0.3s ease;
    width: 10em;

    &:first-child {
      -moz-transform: rotate(0deg) skew(30deg);
      -ms-transform: rotate(0deg) skew(30deg);
      -webkit-transform: rotate(0deg) skew(30deg);
      transform: rotate(0deg) skew(30deg);
    }

    &:nth-child(2) {
      -moz-transform: rotate(60deg) skew(30deg);
      -ms-transform: rotate(60deg) skew(30deg);
      -webkit-transform: rotate(60deg) skew(30deg);
      transform: rotate(60deg) skew(30deg);
    }

    &:nth-child(3) {
      -moz-transform: rotate(120deg) skew(30deg);
      -ms-transform: rotate(120deg) skew(30deg);
      -webkit-transform: rotate(120deg) skew(30deg);
      transform: rotate(120deg) skew(30deg);
    }

    &:nth-child(odd) {
      a {
        background-color: #5691ff;
      }
    }

    &:nth-child(even) {
      a {
        background-color: #83a9ef;
      }
    }

    a {
      -moz-transform: skew(-30deg) rotate(-60deg) scale(1);
      -moz-transition: opacity 0.3s, color 0.3s;
      -ms-transform: skew(-30deg) rotate(-60deg) scale(1);
      -webkit-backface-visibility: hidden;
      -webkit-transform: skew(-30deg) rotate(-60deg) scale(1);
      -webkit-transition: opacity 0.3s, color 0.3s;
      border-radius: 50%;
      bottom: -7.25em;
      color: #fff;
      display: block;
      font-size: 1.18em;
      height: 14.5em;
      padding-top: 1.8em;
      position: absolute;
      right: -7.25em;
      text-align: center;
      text-decoration: none;
      transform: skew(-30deg) rotate(-60deg) scale(1);
      transition: opacity 0.3s, color 0.3s;
      width: 14.5em;
    }
  }
}

.opened-nav {
  -moz-transform: scale(1);
  -ms-transform: scale(1);
  -webkit-transform: scale(1);
  border-radius: 50%;
  pointer-events: auto;
  transform: scale(1);
}

.cn-overlay {
  -moz-transition: all 0.3s ease;
  -webkit-transition: all 0.3s ease;
  background-color: rgba(0, 100, 200, 0.6);
  bottom: 0;
  height: 100%;
  left: 0;
  opacity: 0;
  position: fixed;
  right: 0;
  top: 0;
  transition: all 0.3s ease;
  visibility: hidden;
  width: 100%;
  z-index: 2;

  &.on-overlay {
    opacity: 1;
    visibility: visible;
  }
}

.cn-button {
  background-color: #fff;
  border-radius: 50%;
  border: none;
  bottom: 0;
  box-shadow: 0px 0px 30px -8px #777;
  color: #0a56fc;
  cursor: pointer;
  font-size: 5em;
  height: 5em;
  left: 50%;
  outline: none;
  padding: 1em 0;
  position: fixed;
  text-align: center;
  transform: translate(-50%, 50%);
  width: 5em;
  z-index: 11;

  &:hover,
  &:active,
  &:focus {
    color: #2b6eff;
  }
}

.cn-wrapper li a span {
  font-size: 1.1em;
  opacity: 0.7;
}

.cn-wrapper li.active a {
  background-color: #0a56fc;
}

.cn-wrapper li:not(.active) a:hover,
.cn-wrapper li:not(.active) a:active,
.cn-wrapper li:not(.active) a:focus {
  background-color: #0a56fc;
}
.cn-wrapper li:not(.active) a:focus {
  position: fixed;
}

@media screen and (max-width: 480px) {
  .cn-wrapper {
    font-size: 0.68em;
  }

  .cn-button {
    font-size: 1em;
  }

  .cn-wrapper li {
    font-size: 1.52em;
  }
}
