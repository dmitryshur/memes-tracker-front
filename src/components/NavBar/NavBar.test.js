import React from 'react';
import { shallow } from 'enzyme';
import { NavBar } from './NavBar';

const delay = ms => new Promise(resolve => setTimeout(resolve, ms));

test('testing NavBar when closed', () => {
  const wrapper = shallow(<NavBar navbarStatus="closed" />);
  expect(wrapper.text()).toBe('+');
  expect(wrapper.find('#cn-wrapper').hasClass('opened-nav')).toBe(false);
  expect(wrapper.find('#cn-overlay').hasClass('on-overlay')).toBe(false);
  expect(wrapper).toMatchSnapshot();
});

test('testing NavBar when opened', () => {
  const wrapper = shallow(<NavBar navbarStatus="opened" />);
  expect(wrapper.text()).toBe('-');
  expect(wrapper.find('#cn-wrapper').hasClass('opened-nav')).toBe(true);
  expect(wrapper.find('#cn-overlay').hasClass('on-overlay')).toBe(true);
  expect(wrapper).toMatchSnapshot();
});

test("testing NavBar click when it's initially closed", () => {
  const toggleNavbar = jest.fn();
  const wrapper = shallow(<NavBar navbarStatus="closed" toggleNavbar={toggleNavbar} />);

  Promise.resolve()
    .then(() => {
      wrapper.find('#cn-button').simulate('click', { stopPropagation: () => {} });
      return delay(3000);
    })
    .then(() => {
      expect(toggleNavbar).toHaveBeenCalled();
      expect(wrapper.text()).toBe('-');
      expect(wrapper.find('#cn-wrapper').hasClass('opened-nav')).toBe(false);
      expect(wrapper.find('#cn-overlay').hasClass('on-overlay')).toBe(false);
      expect(wrapper).toMatchSnapshot();
    });
});

test("testing NavBar click when it's opened", () => {
  const toggleNavbar = jest.fn();
  const wrapper = shallow(<NavBar navbarStatus="opened" toggleNavbar={toggleNavbar} />);

  Promise.resolve()
    .then(() => {
      wrapper.find('#cn-button').simulate('click', { stopPropagation: () => {} });
      return delay(3000);
    })
    .then(() => {
      expect(toggleNavbar).toHaveBeenCalled();
      expect(wrapper.text()).toBe('+');
      expect(wrapper.find('#cn-wrapper').hasClass('opened-nav')).toBe(true);
      expect(wrapper.find('#cn-overlay').hasClass('on-overlay')).toBe(true);
      expect(wrapper).toMatchSnapshot();
    });
});
