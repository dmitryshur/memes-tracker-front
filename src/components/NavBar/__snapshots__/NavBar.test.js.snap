// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing NavBar when closed 1`] = `
<React.Fragment>
  <div
    className="cn-button"
    id="cn-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    +
  </div>
  <div
    className="cn-wrapper"
    id="cn-wrapper"
  >
    <ul>
      <li>
        <a
          href="#"
        >
          <span
            className="fas fa-user-plus"
          />
        </a>
      </li>
      <li>
        <a
          href="#"
        >
          <span
            className="fas fa-home"
          />
        </a>
      </li>
      <li>
        <a
          href="#"
        >
          <span
            className="fas fa-star"
          />
        </a>
      </li>
    </ul>
  </div>
  <div
    className="cn-overlay"
    id="cn-overlay"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  />
</React.Fragment>
`;

exports[`testing NavBar when opened 1`] = `
<React.Fragment>
  <div
    className="cn-button"
    id="cn-button"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    -
  </div>
  <div
    className="cn-wrapper opened-nav"
    id="cn-wrapper"
  >
    <ul>
      <li>
        <a
          href="#"
        >
          <span
            className="fas fa-user-plus"
          />
        </a>
      </li>
      <li>
        <a
          href="#"
        >
          <span
            className="fas fa-home"
          />
        </a>
      </li>
      <li>
        <a
          href="#"
        >
          <span
            className="fas fa-star"
          />
        </a>
      </li>
    </ul>
  </div>
  <div
    className="cn-overlay on-overlay"
    id="cn-overlay"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  />
</React.Fragment>
`;
